### PR TITLE
Test/quickcheck state machine

### PR DESCRIPTION
**Pull request by tdietert**
Opening this PR so that I can use the quickcheck state machine tests to observe a potential bug causing stale reads.

### CODE REVIEW

File: test/QuickCheckStateMachine.hs [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+      return (SpawnFailed port1)
+semantics ((_sin, _sout), log) (KillNode (_port, ph)) = do
+  hPutStrLn log "Killing node"
+  terminateProcess (opaque ph)
+  return Ack
+semantics hs (Set i) = do
+  _resp <- command hs ("set x " ++ show i)
+  return Ack
+semantics hs Read = do
+  -- threadDelay 1000000
```
Comment (stevana) [inline test/QuickCheckStateMachine.hs]: This is where I inserted `threadDelay 2000000` earlier and could still reproduce the problem with `test2`.

Comment (tdietert) [reply test/QuickCheckStateMachine.hs]: :+1:

**Comment by stevana** [comment]
Would you be fine with stop using `tasty-discover` and switching over to manually building the tasty test tree?

Because I don't see how I could change the state machine tests to make QuickCheck's default 100 iterations run fast enough for it to be acceptable to run as part of CI, and I also don't know of a way to use `tasty-discover` and be able to change the amount of iterations to run on a per property basis...

**Comment by gdijkstra** [comment]
> I also don't know of a way to use `tasty-discover` and be able to change the amount of iterations to run on a per property basis...

You can use QuickCheck's `withMaxSuccess` to do that per property: https://hackage.haskell.org/package/QuickCheck-2.12.6.1/docs/Test-QuickCheck.html#v:withMaxSuccess

**Comment by stevana** [comment]
Ah cool, thanks!

**Comment by stevana** [comment]
How can I check if this passes CI?

**Comment by tdietert** [comment]
I just enabled "Build forked pull requests" option on CircleCI, so you'll need to push another commit to this branch to get it to build.

**Comment by stevana** [comment]
Cool, seems to work. I think this would be ready for review then.

**Comment by tdietert** [comment]
Looks good to me! Thanks again for all your help reporting a few critical bugs in the protocol.

It's a bit sad we can only have 5 `prop_sequential` tests due to the tests taking ~5s each on average, but I don't think there's any way around it. We could increment it to 10 such that the quickcheck state machine tests would take ~1min :man_shrugging: I'm not sure what's best at this point.

**Pull request merged**